### ISSUE TITLE
kdc/kstash mkey file defaults are inconsistent, and/or documentation is confusing and/or wrong

### ISSUE THREAD

**Issue by riastradh** [heimdal/heimdal#1113]
**Describe the bug**
The krb5.conf(5) man page says:

```
           [kdc]

                      database = {
...
                                 mkey_file = FILENAME
                                      Use this keytab file for the master key
                                      of this database.  If not specified
                                      DATABASENAME.mkey will be used.
```

The kstash(8) man page says:

```
FILES
     /var/heimdal/m-key is the default keyfile if no other keyfile is
     specified.  [...]
```

The kdc(8) man page is silent on the matter.

The info manual is unclear on the connection between krb5.conf and kstash:

```
4.2 Creating the database
...
These defaults can be overriden in the 'database' key in the `kdc'
section of the configuration.

     [kdc]
             database = {
                     dbname = lmdb:/path/to/db-file
                     realm = REALM
                     acl_file = /path/to/kadmind.acl
                     mkey_file = /path/to/mkey
                     log_file = /path/to/iprop-log-file
             }
...
The keys of all the principals are stored in the database.  If you
choose to, these can be encrypted with a master key.  You do not have to
remember this key (or password), but just to enter it once and it will
be stored in a file (`/var/heimdal/m-key').  If you want to have a
master key, run `kstash' to create this master key:
```

Trying to set up a KDC with a minimum of configuration:

1. I created a krb5.conf with:

   ```
   [kdc]
   	database = {
   		dbname = sqlite:/var/heimdal/heimdal.db
   	}
   ```

2. I ran kstash(8), and it created `/var/heimdal/m-key`.

3. I started kdc(8), and it looked for the file `sqlite:/var/heimdal/heimdal.mkey` (note: not `/var/heimdal/heimdal.db.mkey` or even `/var/heimdal/heimdal.mkey`; it stripped off the `.db` suffix, but it actually tried to open a file at a pathname starting with `sqlite:`, as I determined with ktrace).

I suspect the result is that, even though I ran kstash(8), kdc(8) isn't actually using the master key and heimdal.db is unencrypted.  (This doesn't make a real security difference if they're stored on the file system adjacent to one another, but it is confusing.)

**To Reproduce**
As above, with Heimdal 7.7.0.  (If this has changed substantively in master, let me know and I can re-test, but the man pages and info manual at least appear to be unchanged on the matter.)

**Expected behavior**
1. krb5.conf(5), kstash(8), and kdc(8) man pages make it clear what the defaults are.
2. kstash(8) and kdc(8) agree on what defaults are.
3. kdc(8) doesn't take the database type as part of the pathname in the default mkey.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: NetBSD
 - Browser: N/A
 - Version: 10.0_BETA

**Comment by nicowilliams** [comment]
By the way, thank you so much for all these high-quality reports!

**Comment by nicowilliams** [comment]
> * Is there a way to query whether the KDC database, or any individual record in it, is encrypted with an m-key?

There is not.

Before I go any further, we should discuss the value of encrypting any of the HDB with the current scheme or any other scheme.  This applies to MIT as much as Heimdal since both have roughly the same approach to this.

First, as to the current scheme, it seems to have been motivated by the state of the world in the 90s, when it was fairly common to have external disks attached to servers, and so the HDB/KDB key encryption scheme was intended to mitigate theft of those disks from KDCs in data centers.

Second, the current scheme does not bind keys to names or otherwise protect the integrity of any data in an HDB/KDB record other than the keys themselves.  This means that an attack on the system by the disks would be to rename a principal whose keys the attacker knows to a `krbtgt` principal.

Third, the current state of the world is very different from the 90s; the design of HDB/KDB encryption is stale.

My advice: don't bother with HDB/KDB encryption using a master key.  Instead take other steps to secure your KDCs, and do encrypt at the _filesystem_ level (and, if you can get it, _also_ FDE at the hardware level).

> * Is there a way to rewrite a KDC database in place encrypted with an m-key?

Yes: set a master key and then change the principal's keys.

> * If yes, can it be done incrementally, in case there is a power outage in the middle?

Yes.  Each principal's keys may or may not be encrypted -- if that principal's keys are encrypted, then all of that principal's keys are encrypted, not just some.

> * Is there a way to make `kadmin add` and similar noisily fail to create a record if there is no m-key available?

There is not, not currently.

I'd rather rip out HDB encryption and replace it with advice on how to secure KDCs.

**Comment by nicowilliams** [comment]
> > I'd rather rip out HDB encryption and replace it with advice on how to secure KDCs.
> 
> That sounds reasonable, but then that presents a complementary question: Suppose I have already used an mkey. Can I convert the database to stop using it, so I can remove the mkey stuff from the configuration?

Indeed, I can't just rip it out, though I can deprecate it and document that and make it harder to make new uses of mkeys.

> Can I convert the database to stop using it, so I can remove the mkey stuff from the configuration?

Yes, I think so.  I think you'd `kadmin -l dump --decrypt` then `kadmin load`.
